In [1]:
#import libraries
import pandas as pd
import datetime as dt

import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
#import dataset
data = pd.read_csv('online_retail.csv')

### Explore the data

In [3]:
data.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

           InvoiceDate  Price  Customer ID         Country  
0  2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1  2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2  2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3  2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4  2009-12-01 07:45:00   1.25      13085.0  United Kingdom

#### Feature description
- Invoice: A 6-digit systematically assigned unique code for transactions. This makes it easier to take payments and overdues.
- StockCode: Number uniquely assigned to each product.
- Description: Name of the product explained in a detail manner. 
- Quantity: The quantity of product bought for each transaction.
- InvoiceDate: Date and time of the transaction
- Price: Cost of each unit of the product.
- CustomerID: Unique id assigned to each user.
- Country: Name of the country where the customer/user resides.

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  object 
 1   StockCode    1067371 non-null  object 
 2   Description  1062989 non-null  object 
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  object 
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 65.1+ MB


In [5]:
Arr_C = data[data['Invoice'].str.contains('^[a-zA-Z]+', regex=True)]['Invoice'].unique()
Arr_C

array(['C489449', 'C489459', 'C489476', ..., 'C581499', 'C581568',
       'C581569'], dtype=object)

- Invoce number has details regarding the transactions. And few transactions has alpabet C assigned which corresponds to the cancelled order

In [6]:
#column to show if the row belongs to cancelled order or not
data['Cancelled'] = data['Invoice'].apply(lambda x:int('C' in x))
data['Cancelled'].value_counts()

0    1047877
1      19494
Name: Cancelled, dtype: int64

- Approximately 1.8% of the total data is regarding cancelled orders

In [7]:
#there are a few values of stock code with alphabets, analyze them
codes = data[data['StockCode'].str.contains('^[a-zA-Z]+', regex=True)]['StockCode'].unique()
codes

array(['POST', 'D', 'DCGS0058', 'DCGS0068', 'DOT', 'M', 'DCGS0004',
       'DCGS0076', 'C2', 'BANK CHARGES', 'DCGS0003', 'TEST001',
       'gift_0001_80', 'DCGS0072', 'gift_0001_20', 'DCGS0044', 'TEST002',
       'gift_0001_10', 'gift_0001_50', 'DCGS0066N', 'gift_0001_30',
       'PADS', 'ADJUST', 'gift_0001_40', 'gift_0001_60', 'gift_0001_70',
       'gift_0001_90', 'DCGSSGIRL', 'DCGS0006', 'DCGS0016', 'DCGS0027',
       'DCGS0036', 'DCGS0039', 'DCGS0060', 'DCGS0056', 'DCGS0059', 'GIFT',
       'DCGSLBOY', 'm', 'DCGS0053', 'DCGS0062', 'DCGS0037', 'DCGSSBOY',
       'DCGSLGIRL', 'S', 'DCGS0069', 'DCGS0070', 'DCGS0075', 'B',
       'DCGS0041', 'ADJUST2', 'C3', 'SP1002', 'AMAZONFEE', 'DCGS0055',
       'DCGS0074', 'DCGS0057', 'DCGS0073', 'DCGS0071', 'DCGS0066P',
       'DCGS0067', 'CRUK'], dtype=object)

In [8]:
#Try to understand what the code means by going through the corresponding descriptions
for code in codes:
    print("{:<15} -> {:<30}".format(code, data[data['StockCode'] == code]['Description'].unique()[0]))

POST            -> POSTAGE                       
D               -> Discount                      
DCGS0058        -> MISO PRETTY  GUM              
DCGS0068        -> DOGS NIGHT COLLAR             
DOT             -> DOTCOM POSTAGE                
M               -> Manual                        
DCGS0004        -> HAYNES CAMPER SHOULDER BAG    
DCGS0076        -> SUNJAR LED NIGHT NIGHT LIGHT  
C2              -> CARRIAGE                      
BANK CHARGES    ->  Bank Charges                 
DCGS0003        -> BOXED GLASS ASHTRAY           
TEST001         -> This is a test product.       
gift_0001_80    -> nan                           
DCGS0072        -> CAT CAMOUFLAGUE COLLAR        
gift_0001_20    -> Dotcomgiftshop Gift Voucher £20.00
DCGS0044        -> HANDZ-OFF CAR FRESHENER       
TEST002         -> This is a test product.       
gift_0001_10    -> Dotcomgiftshop Gift Voucher £10.00
gift_0001_50    -> Dotcomgiftshop Gift Voucher £50.00
DCGS0066N       -> NAVY CUDDLES DOG HO

- Summary: What the code means like transactions with gift card, discount, amazon fee or for testing is understood from the description. Some values are missing and the others are understood. 

In [9]:
data['Country'].value_counts()

United Kingdom          981330
EIRE                     17866
Germany                  17624
France                   14330
Netherlands               5140
Spain                     3811
Switzerland               3189
Belgium                   3123
Portugal                  2620
Australia                 1913
Channel Islands           1664
Italy                     1534
Norway                    1455
Sweden                    1364
Cyprus                    1176
Finland                   1049
Austria                    938
Denmark                    817
Unspecified                756
Greece                     663
Japan                      582
Poland                     535
USA                        535
United Arab Emirates       500
Israel                     371
Hong Kong                  364
Singapore                  346
Malta                      299
Iceland                    253
Canada                     228
Lithuania                  189
RSA                        169
Bahrain 

- List of countries from where the transactions were made is shown. United kingdom has the maximum orders made and Saudi Arabia has the minimum.

In [10]:
data.describe()

Quantity         Price    Customer ID     Cancelled
count  1.067371e+06  1.067371e+06  824364.000000  1.067371e+06
mean   9.938898e+00  4.649388e+00   15324.638504  1.826357e-02
std    1.727058e+02  1.235531e+02    1697.464450  1.339030e-01
min   -8.099500e+04 -5.359436e+04   12346.000000  0.000000e+00
25%    1.000000e+00  1.250000e+00   13975.000000  0.000000e+00
50%    3.000000e+00  2.100000e+00   15255.000000  0.000000e+00
75%    1.000000e+01  4.150000e+00   16797.000000  0.000000e+00
max    8.099500e+04  3.897000e+04   18287.000000  1.000000e+00

- Quantity and price have negative minimum values because of the cancelled orders. 

In [11]:
data.isnull().sum()

Invoice             0
StockCode           0
Description      4382
Quantity            0
InvoiceDate         0
Price               0
Customer ID    243007
Country             0
Cancelled           0
dtype: int64

- There are few null or missing values in description and customer ID. So delete the duplicate rows and the ones with missing values

#### Clean the data

In [12]:
data.duplicated().sum()

34335

In [13]:
# drop duplicates
data.drop_duplicates(inplace=True)

In [14]:
# drop the rows with null values customer ID
data = data[(data['Customer ID'].notnull())]

In [15]:
# delete rows corresponding to the cancelled orders
data = data[(data['Price']>0) & (data['Quantity']>0)]

In [16]:
#drop the cancelled column as all transactions corresponding to cancelled orders are deleted
data = data.drop('Cancelled',axis=True)

In [17]:
data

Invoice StockCode                          Description  Quantity  \
0        489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1        489434    79323P                   PINK CHERRY LIGHTS        12   
2        489434    79323W                  WHITE CHERRY LIGHTS        12   
3        489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4        489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...         ...       ...                                  ...       ...   
1067366  581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
1067367  581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
1067368  581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
1067369  581587     22138        BAKING SET 9 PIECE RETROSPOT          3   
1067370  581587      POST                              POSTAGE         1   

                 InvoiceDate  Price  Customer ID         Country  
0        2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1        2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2        2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3        2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4        2009-12-01 07:45:00   1.25      13085.0  United Kingdom  
...                      ...    ...          ...             ...  
1067366  2011-12-09 12:50:00   2.10      12680.0          France  
1067367  2011-12-09 12:50:00   4.15      12680.0          France  
1067368  2011-12-09 12:50:00   4.15      12680.0          France  
1067369  2011-12-09 12:50:00   4.95      12680.0          France  
1067370  2011-12-09 12:50:00  18.00      12680.0          France  

[779425 rows x 8 columns]

### Recency, Frequency, Monetary Analysis

#### RFM Calculation
- recency: when was the last order of a customer
- frequency: purchases made in a particular period of time
- momentary: total amunt spend in the given period

In [18]:
#add a column showing the total amount of each transaction
data['T Price'] = data['Quantity']*data['Price']

In [19]:
#convert InvoiceDate from string to datetime
data['InvoiceDate']  = pd.to_datetime(data['InvoiceDate'])

In [20]:
#Coulmn showing only the date of transaction excluding the time from invoice date column
data['Date'] = data['InvoiceDate'].apply(lambda x:dt.datetime(x.year,x.month,x.day))

In [21]:
#as the data is from 2009-10 for later calculations the last date of transaction is taken as maximum instead of today's date
max_date = data['Date'].max() + dt.timedelta(days=1)
max_date

Timestamp('2011-12-10 00:00:00')

In [22]:
#create a new dataframe with the recency, frequency and monetary values as it's columns
rfm = data.groupby(['Customer ID']).agg({'InvoiceDate':lambda x:(max_date - x.max()).days,'Invoice':'count','T Price': 'sum'})

In [23]:
rfm.rename(columns={'InvoiceDate':'Recency','Invoice':'Frequency','T Price':'MonetaryValue'},inplace= True)

In [24]:
rfm

Recency  Frequency  MonetaryValue
Customer ID                                   
12346.0          325         34       77556.46
12347.0            2        222        4921.53
12348.0           75         51        2019.40
12349.0           18        175        4428.69
12350.0          310         17         334.40
...              ...        ...            ...
18283.0            3        938        2664.90
18284.0          431         28         461.68
18285.0          660         12         427.00
18286.0          476         67        1296.43
18287.0           42        155        4182.99

[5878 rows x 3 columns]

- assigning scores using quatiles

In [25]:
#Scores from 1 to 4 are given to customers of equal bins based on their recency
#Customers who bought something very recently are given a score 4 and ones who didn't buy anything for a long time are scored 1
r_labels = range(4,0,-1)
rfm['r_score'] = pd.qcut(rfm.Recency,q=4,labels=r_labels)

#Those who bought less frequnetly are scored 1 and more frequent ones are given 4
f_labels = range(1,5,1)
rfm['f_score'] = pd.qcut(rfm.Frequency,q=4,labels=f_labels)

#Those who spent very less money are given 1 and the ones that spent the most are given 5
m_labels = range(1,5,1)
rfm['m_score'] = pd.qcut(rfm.MonetaryValue,q=4,labels=m_labels)

rfm

Recency  Frequency  MonetaryValue r_score f_score m_score
Customer ID                                                           
12346.0          325         34       77556.46       2       2       4
12347.0            2        222        4921.53       4       4       4
12348.0           75         51        2019.40       3       2       3
12349.0           18        175        4428.69       4       4       4
12350.0          310         17         334.40       2       1       1
...              ...        ...            ...     ...     ...     ...
18283.0            3        938        2664.90       4       4       4
18284.0          431         28         461.68       1       2       2
18285.0          660         12         427.00       1       1       2
18286.0          476         67        1296.43       1       3       3
18287.0           42        155        4182.99       3       4       4

[5878 rows x 6 columns]

In [26]:
rfm.to_csv('RFM.csv',index=True)

In [28]:
#calculate the concatenated score and the total score from r,f,m scores
rfm['conc_score'] = rfm.apply(lambda x:str(x['r_score'])+str(x['f_score'])+str(x['m_score']), axis = 1) #ranges from 111 to 444
rfm['sum_score'] = rfm[['r_score','f_score','m_score']].sum(axis=1) #range from 3 to 12

In [29]:
rfm

Recency  Frequency  MonetaryValue r_score f_score m_score  \
Customer ID                                                              
12346.0          325         34       77556.46       2       2       4   
12347.0            2        222        4921.53       4       4       4   
12348.0           75         51        2019.40       3       2       3   
12349.0           18        175        4428.69       4       4       4   
12350.0          310         17         334.40       2       1       1   
...              ...        ...            ...     ...     ...     ...   
18283.0            3        938        2664.90       4       4       4   
18284.0          431         28         461.68       1       2       2   
18285.0          660         12         427.00       1       1       2   
18286.0          476         67        1296.43       1       3       3   
18287.0           42        155        4182.99       3       4       4   

            conc_score  sum_score  
Customer ID                        
12346.0            224          8  
12347.0            444         12  
12348.0            323          8  
12349.0            444         12  
12350.0            211          4  
...                ...        ...  
18283.0            444         12  
18284.0            122          5  
18285.0            112          4  
18286.0            133          7  
18287.0            344         11  

[5878 rows x 8 columns]

In [30]:
#summarize 
#group customers based on sum score and calcuate the mean of r,f,m in each group along with the number of customers in each group
rfm_agg = rfm.groupby('sum_score').agg({'Recency' : 'mean','Frequency' : 'mean','MonetaryValue' : ['mean', 'count']})
rfm_agg

Recency   Frequency MonetaryValue      
                 mean        mean          mean count
sum_score                                            
3          544.872495    9.438980    172.558233   549
4          380.410811   17.189189    247.059407   555
5          314.886762   24.515152    425.468900   627
6          232.686495   34.332797    687.600498   622
7          193.354067   50.814992   1098.270115   627
8          146.298387   72.861290   1452.786731   620
9           95.640429  109.402504   2306.293567   559
10          66.966192  169.322064   3165.090411   562
11          35.702602  274.312268   4993.429541   538
12           9.730210  560.657512  14708.704643   619

- Examining the mean values of recency, frequency and monetary values for each of the RFM scores from 3 to 12 gives a better picture of the customers. Their are total of 549 customers in the lowest scoring group with a score of 3, in the rfm segment 111 havhaving a mean recency of 545 days, placed an average of 9 orders and spent an average of £172.5. Top customers with score 15 and the rfm segment 555 have been seen 10 days ago, have placed an average of 560 orders and spent £14708.7

In [31]:
#group the customers using rfm score
score_labels = ['Gold','Silver','Bronze']
score_groups = pd.qcut(rfm.sum_score, q = 3, labels = score_labels)
rfm['RFM Group'] = score_groups.values
rfm.head()

Recency  Frequency  MonetaryValue r_score f_score m_score  \
Customer ID                                                              
12346.0          325         34       77556.46       2       2       4   
12347.0            2        222        4921.53       4       4       4   
12348.0           75         51        2019.40       3       2       3   
12349.0           18        175        4428.69       4       4       4   
12350.0          310         17         334.40       2       1       1   

            conc_score  sum_score RFM Group  
Customer ID                                  
12346.0            224          8    Silver  
12347.0            444         12    Bronze  
12348.0            323          8    Silver  
12349.0            444         12    Bronze  
12350.0            211          4      Gold

In [32]:
#summarize the gold, silver, bronze groups
group_summ = rfm.groupby('RFM Group').agg({'Recency' : 'mean','Frequency' : 'mean','MonetaryValue' : ['mean', 'count']})
group_summ

Recency   Frequency MonetaryValue      
                 mean        mean          mean count
RFM Group                                            
Gold       362.272843   21.864853    393.671462  2353
Silver     146.955150   76.517719   1593.887729  1806
Bronze      36.571262  343.098313   7894.086142  1719